Dart/src/com/jetbrains/lang/dart/ide/editor/DartEnterInStringHandler.java

```diff
@@ -31,37 +31,30 @@ public Result preprocessEnter(@NotNull final PsiFile file,
     Document document = editor.getDocument();
     ASTNode token = psiAtOffset.getNode();
     IElementType type = token.getElementType();
+    if (type == DartTokenTypes.RAW_TRIPLE_QUOTED_STRING) {
+      return Result.DefaultSkipIndent; // Multiline string gets no indent
+    }
     if (type == DartTokenTypes.RAW_SINGLE_QUOTED_STRING) {
       char quote = token.getText().charAt(1);
       breakString("r" + quote, String.valueOf(quote), caretOffsetRef, caretAdvanceRef, document);
       return Result.Default;
     }
-    if (type == DartTokenTypes.REGULAR_STRING_PART) {
-      token = token.getTreeNext();
-      if (token == null) return Result.Continue; // Can't happen with current grammar.
+    while ((token = token.getTreeParent()) != null) {
```

Comment:
I think we have to do something. The current code is buggy when trying to
split embedded interpolations.

var b = "see ${'${a}'} and more";

If you type a Return when the cursor is before the 'a' in '{a}' you get
incorrect code. It should insert a newline and indent but no additional
quotes.

On Thu, Sep 10, 2015 at 9:52 AM, Alexander Doroshko <
notifications@github.com> wrote:

> In
> Dart/src/com/jetbrains/lang/dart/ide/editor/DartEnterInStringHandler.java
> <https://github.com/JetBrains/intellij-plugins/pull/255#discussion_r39183494>
> :
>
> >      if (type == DartTokenTypes.RAW_SINGLE_QUOTED_STRING) {
> >        char quote = token.getText().charAt(1);
> >        breakString("r" + quote, String.valueOf(quote), caretOffsetRef, caretAdvanceRef, document);
> >        return Result.Default;
> >      }
> > -    if (type == DartTokenTypes.REGULAR_STRING_PART) {
> > -      token = token.getTreeNext();
> > -      if (token == null) return Result.Continue; // Can't happen with current grammar.
> > +    while ((token = token.getTreeParent()) != null) {
>
> Do you think we can optimize it by checking not more than 2 parents up the
> hierarchy?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JetBrains/intellij-plugins/pull/255/files#r39183494>.
>